File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    void dfs(vector<vector<char>>& grid, int r, int c){
        int row=grid.size();
        int col=grid[0].size();
        
        if(r<0 || r>=row || c<0 || c>=col || grid[r][c]!='1'){
            return;
        }
        vector<int> rNbr = {-1,0,1,0};
        vector<int> cNbr = {0,1,0,-1};
        grid[r][c]=0;
        for(int i=0;i<4;i++){
            int nrow=r+rNbr[i];
            int ncol=c+cNbr[i];
            dfs(grid,nrow,ncol);
        }

    }
    int numIslands(vector<vector<char>>& grid) {
        int n=grid.size();
        int m=grid[0].size();
        int cnt=0;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]=='1'){
                    cnt++;
                    dfs(grid,i,j);
                }
            }
        }
        return cnt;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public void dfs(int r,int c,char[][] grid,boolean[][] vis){
        int n=grid.length;
        int m=grid[0].length;
        

        if(r<0 || c<0 || r>=n || c>=m || vis[r][c] ||grid[r][c] == '0'){
            return;
        }
        vis[r][c]=true;

        dfs(r-1,c,grid,vis);
        dfs(r,c+1,grid,vis);
        dfs(r+1,c,grid,vis);
        dfs(r,c-1,grid,vis);

    }
    public int numIslands(char[][] grid) {
        int n=grid.length;
        int m=grid[0].length;
        int ans=0;

        boolean[][] vis=new boolean[n][m];

        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(!vis[i][j] && grid[i][j]=='1'){
                    dfs(i,j,grid,vis);
                    ans++;
                }
            }
        }

        return ans;

    }
}